Experiment with Gutenberg from Bintray

mokagio opened the pull request:
This PR is not meant for merging. Once the setup works, we'll create a clean new one.

Expect lots of notes to self in the comments. I'm also trying to document every experiment and failure in the commit messages.

mokagio committed:
Add configurations for Gutenberg Mobile Bintray repo

mokagio committed:
Add placeholder code for configuring Gutenberg as a local build

mokagio committed:
Add Gutenberg dependency from Bintray – Fails

./gradlew assembleDebug fails with

> Task :@wordpress_react-native-bridge:bundleDebugAar FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':@wordpress_react-native-bridge:bundleDebugAar'.
> Direct local .aar file dependencies are not supported when building an AAR. The resulting AAR would be broken because the classes and Android resources from any local .aar file dependencies would not be packaged in the resulting AAR. Previous versions of the Android Gradle Plugin produce broken AARs in this case too (despite not throwing this error). The following direct local .aar file dependencies of the :@wordpress_react-native-bridge project caused this error: /Users/gio/Developer/a8c/wpandroid/libs/gutenberg-mobile/gutenberg/packages/react-native-bridge/android/vendor/hermes-engine/hermes-debug.aar

Might it be due to the sort of double import of Gutenberg? One from
Bintray the other from the local submodule?

mokagio committed:
Add mokagio's maven repo for the Gutenberg AAR to lib/editor

mokagio committed:
Switch to Gutenberg via Bintray for WordPressEditor – Fails

For some reason, it looks for the library only on Jitpack, where it fails with
a 401, not in the my (mokagio) maven repo

mokagio committed:
Try to remove jitpack – Still fails, now it can't find Aztec either

I though maybe if I get rid of jitpack it'll use my repo for Gutenmberg
and be happy. Didn't consider the dependency on Aztec. I guess I need to
find a way to tell WordPressEditor to look for it into the Gutenmberg AAR.

Failed with

FAILURE: Build failed with an exception.

* What went wrong:
Could not determine the dependencies of task ':libs:editor:WordPressEditor:compileDebugAidl'.
> Could not resolve all task dependencies for configuration ':libs:editor:WordPressEditor:debugCompileClasspath'.
   > Could not find com.github.wordpress-mobile.WordPress-Aztec-Android:aztec:v1.3.45.
     Required by:
         project :libs:editor:WordPressEditor
   > Could not find com.github.wordpress-mobile.WordPress-Aztec-Android:wordpress-shortcodes:v1.3.45.
     Required by:
         project :libs:editor:WordPressEditor
   > Could not find com.github.wordpress-mobile.WordPress-Aztec-Android:wordpress-comments:v1.3.45.
     Required by:
         project :libs:editor:WordPressEditor
   > Could not find com.github.wordpress-mobile.gutenberg-mobiled:react-native-bridge:20201210-144401.
     Required by:
         project :libs:editor:WordPressEditor

mokagio committed:
Fix WordPressEditor's new Gutenberg dependency path -.-' re-add Jitpack

All those "can't find..." were because I mistyped the project name...

---

Now it fails with:

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':@wordpress_react-native-bridge:bundleDebugAar'.
> Direct local .aar file dependencies are not supported when building an AAR. The resulting AAR would be broken because the classes and Android resources from any local .aar file dependencies would not be packaged in the resulting AAR. Previous versions of the Android Gradle Plugin produce broken AARs in this case too (despite not throwing this error). The following direct local .aar file dependencies of the :@wordpress_react-native-bridge project caused this error: /Users/gio/Developer/a8c/wpandroid/libs/gutenberg-mobile/gutenberg/packages/react-native-bridge/android/vendor/hermes-engine/hermes-debug.aar

I'm guessing that's because Gradle is trying to get the same Gutenberg from two
locations?

mokagio committed:
Bluntly comment RN configs from settings.gradle – assembleDebug builds! 😱

mokagio commented:
Lint aside, this builds on CI and runs on device.

It's far from done, though. I went in the Gutenber submodule and checked out an old tag as a test see if the build worked and got a null pointer exception when running `./gradlew assembleDebug`.

Clearly, there are still settings pointing to the files in the repo rather than the AAR.

An open question I have is whether the Azetec library that WordPressEditor refers to comes from the fat AAR or somewhere else.

mokagio commented:
Keen for your feedback @mchowning 😄 🙏

mchowning commented on libs/editor/WordPressEditor/build.gradle:
```diff
@@ -79,10 +81,13 @@ dependencies {
-    }
+    // The version value here is the same as the one in the repo's root
+    // build.gradle
+    implementation "com.github.wordpress-mobile.gutenberg-mobile:react-native-gutenberg-bridge:20201210-144401"
```
Can probably use `rootProject.ext.gutenbergMobileVersion` here for the version number.

mchowning commented on settings.gradle:
```diff
@@ -89,5 +90,17 @@ if (localBuilds.exists()) {
     }
+
+    // Not sure what to use in place of :WordPressUtils here. Leaving it for
+    // now.
```
I agree that ignoring this for now makes sense. WPAndroid already has a working capability to build using gb-mobile source by using the `wp.BUILD_GUTENBERG_FROM_SOURCE` flag, so we can focus on getting the binary build working (without breaking the current build-from-source mode). If you get that working you're going to have a lot of new Android BFFs!

After that, we could work on updating our build-from-source to use the new (and better) approach in a separate follow-up PR. With that said, if you want to go ahead and knock them both out, don't let me stop you! 😆

mokagio committed:
Read Gutenberg binary settings from rootProject instead of hardcoding

Thanks @mchowning.

See https://github.com/wordpress-mobile/WordPress-Android/pull/13577#discussion_r540518735

mokagio committed:
Add note regarding BUILD_GUTENBERG_FROM_SOURCE in WordPressEditor

mokagio committed:
Fix FIXME comment 🤦‍♂️

mokagio committed:
Hardcode BUILD_GUTENBERG_FROM_SOURCE to false in WordPressEditor

This is just to test whether the build without the Gutenberg submodule
works.

Answer is: it doesn't. There must still be something that requires the
code in there.

Here's the stacktrace I get when running assembleDebug after reverting
the Gutenberg submodule to v0.1.0

> Task :libs:editor:WordPressEditor:packageDebugAssets FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':libs:editor:WordPressEditor:packageDebugAssets'.
> java.lang.NullPointerException (no error message)

* Try:
Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':libs:editor:WordPressEditor:packageDebugAssets'.
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.lambda$executeIfValid$1(ExecuteActionsTaskExecuter.java:205)
        at org.gradle.internal.Try$Failure.ifSuccessfulOrElse(Try.java:263)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeIfValid(ExecuteActionsTaskExecuter.java:203)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:184)
        at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:109)
        at org.gradle.api.internal.tasks.execution.FinalizePropertiesTaskExecuter.execute(FinalizePropertiesTaskExecuter.java:46)
        at org.gradle.api.internal.tasks.execution.ResolveTaskExecutionModeExecuter.execute(ResolveTaskExecutionModeExecuter.java:62)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:57)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:56)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:36)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:77)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:55)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:52)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:416)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:406)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:165)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:250)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:158)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:102)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.call(DelegatingBuildOperationExecutor.java:36)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:52)
        at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:41)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:372)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:359)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:352)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:338)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.lambda$run$0(DefaultPlanExecutor.java:127)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:191)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.executeNextNode(DefaultPlanExecutor.java:182)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:124)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
Caused by: java.lang.NullPointerException
        at com.android.ide.common.resources.DataSet.handleChangedFile(DataSet.java:572)
        at com.android.ide.common.resources.DataSet.updateWith(DataSet.java:516)
        at com.android.build.gradle.tasks.MergeSourceSetFolders.doIncrementalTaskAction(MergeSourceSetFolders.kt:182)
        at com.android.build.gradle.internal.tasks.IncrementalTask.handleIncrementalInputs(IncrementalTask.kt:112)
        at com.android.build.gradle.internal.tasks.IncrementalTask.access$handleIncrementalInputs(IncrementalTask.kt:64)
        at com.android.build.gradle.internal.tasks.IncrementalTask$taskAction$$inlined$recordTaskAction$1.invoke(AndroidVariantTask.kt:73)
        at com.android.build.gradle.internal.tasks.IncrementalTask$taskAction$$inlined$recordTaskAction$1.invoke(AndroidVariantTask.kt:34)
        at com.android.build.gradle.internal.tasks.Blocks.recordSpan(Blocks.java:91)
        at com.android.build.gradle.internal.tasks.IncrementalTask.taskAction$gradle(IncrementalTask.kt:135)
        at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:104)
        at org.gradle.api.internal.project.taskfactory.IncrementalTaskInputsTaskAction.doExecute(IncrementalTaskInputsTaskAction.java:47)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:42)
        at org.gradle.api.internal.project.taskfactory.AbstractIncrementalTaskAction.execute(AbstractIncrementalTaskAction.java:25)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:28)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$3.run(ExecuteActionsTaskExecuter.java:568)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:402)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:394)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:165)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:250)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:158)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:92)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.run(DelegatingBuildOperationExecutor.java:31)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:553)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:536)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.access$300(ExecuteActionsTaskExecuter.java:109)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$TaskExecution.executeWithPreviousOutputFiles(ExecuteActionsTaskExecuter.java:276)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$TaskExecution.execute(ExecuteActionsTaskExecuter.java:265)
        at org.gradle.internal.execution.steps.ExecuteStep.lambda$execute$0(ExecuteStep.java:32)
        at org.gradle.internal.execution.steps.ExecuteStep.execute(ExecuteStep.java:32)
        at org.gradle.internal.execution.steps.ExecuteStep.execute(ExecuteStep.java:26)
        at org.gradle.internal.execution.steps.CleanupOutputsStep.execute(CleanupOutputsStep.java:63)
        at org.gradle.internal.execution.steps.CleanupOutputsStep.execute(CleanupOutputsStep.java:35)
        at org.gradle.internal.execution.steps.ResolveInputChangesStep.execute(ResolveInputChangesStep.java:49)
        at org.gradle.internal.execution.steps.ResolveInputChangesStep.execute(ResolveInputChangesStep.java:34)
        at org.gradle.internal.execution.steps.CancelExecutionStep.execute(CancelExecutionStep.java:43)
        at org.gradle.internal.execution.steps.TimeoutStep.executeWithoutTimeout(TimeoutStep.java:73)
        at org.gradle.internal.execution.steps.TimeoutStep.execute(TimeoutStep.java:54)
        at org.gradle.internal.execution.steps.CatchExceptionStep.execute(CatchExceptionStep.java:34)
        at org.gradle.internal.execution.steps.CreateOutputsStep.execute(CreateOutputsStep.java:44)
        at org.gradle.internal.execution.steps.SnapshotOutputsStep.execute(SnapshotOutputsStep.java:54)
        at org.gradle.internal.execution.steps.SnapshotOutputsStep.execute(SnapshotOutputsStep.java:38)
        at org.gradle.internal.execution.steps.BroadcastChangingOutputsStep.execute(BroadcastChangingOutputsStep.java:49)
        at org.gradle.internal.execution.steps.CacheStep.executeWithoutCache(CacheStep.java:153)
        at org.gradle.internal.execution.steps.CacheStep.executeAndStoreInCache(CacheStep.java:129)
        at org.gradle.internal.execution.steps.CacheStep.lambda$executeWithCache$2(CacheStep.java:107)
        at org.gradle.internal.execution.steps.CacheStep.lambda$executeWithCache$3(CacheStep.java:107)
        at org.gradle.internal.Try$Success.map(Try.java:162)
        at org.gradle.internal.execution.steps.CacheStep.executeWithCache(CacheStep.java:76)
        at org.gradle.internal.execution.steps.CacheStep.execute(CacheStep.java:66)
        at org.gradle.internal.execution.steps.CacheStep.execute(CacheStep.java:41)
        at org.gradle.internal.execution.steps.StoreExecutionStateStep.execute(StoreExecutionStateStep.java:44)
        at org.gradle.internal.execution.steps.StoreExecutionStateStep.execute(StoreExecutionStateStep.java:33)
        at org.gradle.internal.execution.steps.RecordOutputsStep.execute(RecordOutputsStep.java:38)
        at org.gradle.internal.execution.steps.RecordOutputsStep.execute(RecordOutputsStep.java:24)
        at org.gradle.internal.execution.steps.SkipUpToDateStep.executeBecause(SkipUpToDateStep.java:92)
        at org.gradle.internal.execution.steps.SkipUpToDateStep.lambda$execute$0(SkipUpToDateStep.java:85)
        at org.gradle.internal.execution.steps.SkipUpToDateStep.execute(SkipUpToDateStep.java:55)
        at org.gradle.internal.execution.steps.SkipUpToDateStep.execute(SkipUpToDateStep.java:39)
        at org.gradle.internal.execution.steps.ResolveChangesStep.execute(ResolveChangesStep.java:76)
        at org.gradle.internal.execution.steps.ResolveChangesStep.execute(ResolveChangesStep.java:37)
        at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsFinishedStep.execute(MarkSnapshottingInputsFinishedStep.java:36)
        at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsFinishedStep.execute(MarkSnapshottingInputsFinishedStep.java:26)
        at org.gradle.internal.execution.steps.ResolveCachingStateStep.execute(ResolveCachingStateStep.java:94)
        at org.gradle.internal.execution.steps.ResolveCachingStateStep.execute(ResolveCachingStateStep.java:49)
        at org.gradle.internal.execution.steps.CaptureStateBeforeExecutionStep.execute(CaptureStateBeforeExecutionStep.java:79)
        at org.gradle.internal.execution.steps.CaptureStateBeforeExecutionStep.execute(CaptureStateBeforeExecutionStep.java:53)
        at org.gradle.internal.execution.steps.ValidateStep.execute(ValidateStep.java:74)
        at org.gradle.internal.execution.steps.SkipEmptyWorkStep.lambda$execute$2(SkipEmptyWorkStep.java:78)
        at org.gradle.internal.execution.steps.SkipEmptyWorkStep.execute(SkipEmptyWorkStep.java:78)
        at org.gradle.internal.execution.steps.SkipEmptyWorkStep.execute(SkipEmptyWorkStep.java:34)
        at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsStartedStep.execute(MarkSnapshottingInputsStartedStep.java:39)
        at org.gradle.internal.execution.steps.LoadExecutionStateStep.execute(LoadExecutionStateStep.java:40)
        at org.gradle.internal.execution.steps.LoadExecutionStateStep.execute(LoadExecutionStateStep.java:28)
        at org.gradle.internal.execution.impl.DefaultWorkExecutor.execute(DefaultWorkExecutor.java:33)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeIfValid(ExecuteActionsTaskExecuter.java:192)
        ... 30 more

mokagio committed:
Remove symlink to gutenberg-mobile resource from editor source

This is what was causing the null pointer exception.

I noticed it by looking at the project tree in Android Studio.

mokagio committed:
Remove symlink to supported-blocks.json from WordPress assets

mokagio committed:
Import hermes directly in editor as a temporary crash fix

This reverts commit cf972ee855e82f0613dfe23e46153488cd4fc292.

mokagio committed:
Use compileOnly- for hermes

mchowning commented:
> An open question I have is whether the Azetec library that WordPressEditor refers to comes from the fat AAR or somewhere else.

This is a little tricky since it is a dependency of both gutenberg-mobile and the WordPressEditor. If it could be included in the fat aar and have WordPressEditor "depend" on it via gutenberg-mobile, it would be an easy way to ensure the versions stay in sync. That might require changing some dependencies from `compile` or `implementation` to `api` though.

mokagio committed:
Point RN bridge to new version built including the JS bundle

mokagio committed:
Get the app to run and load the editor as well as assemble with Gradle

mokagio committed:
Only define Hermes dependency in WP, no need for it in editor

mokagio committed:
WIP - Restore option to build Gutenberg from the submodule source

mokagio committed:
Update gutenberg-mobile submodule with WIP Bintray build version

mokagio committed:
Revert me! - Track .idea/codeStyles/Project.xml change

This keeps changing and it's distracting me. Committing it to get it out
of the way.

mchowning commented:
I only had time to do a quick test, but using c397b9e from this PR and 39bb2901be from [the gutenberg PR](https://github.com/WordPress/gutenberg/pull/27836), both the from-source and from-binary builds worked for me! 🎉  This is some great work @mokagio !

mokagio committed:
Update RN bridge library to version with the supported-blocks asset

mokagio committed:
Update RN bridge to version including the unsupported-block-editor files

Also had to update the path in the calls to load those files, because
now they're just in the root of the library.

mokagio commented:
This was an experimental branch where to mess around with the setup but I needed a PR to share the experiments and run the tests in CI. This is done now. Closing.
